client/src/main/java/io/pravega/client/stream/impl/ControllerResolverFactory.java

```diff
@@ -161,26 +158,62 @@ public void shutdown() {
             }
         }
 
+        @Override
+        @Synchronized
+        public void refresh() {
+            if (started.get()) {
+                this.scheduledExecutor.execute(this::getControllers);
+            }
+        }
+
         // The controller discovery API invoker.
         private void getControllers() {
+            log.debug("Attempting to refresh the controller server endpoints");
             try {
-                final ServerResponse controllerServerList =
-                        this.client.getControllerServerList(ServerRequest.getDefaultInstance());
-                final ResolvedServerInfoGroup serverInfoGroup = ResolvedServerInfoGroup.builder()
-                        .addAll(controllerServerList.getNodeURIList()
-                                .stream()
-                                .map(node ->
-                                        new ResolvedServerInfo(
-                                                new InetSocketAddress(node.getEndpoint(), node.getPort())))
-                                .collect(Collectors.toList()))
-                        .build();
+                final ResolvedServerInfoGroup serverInfoGroup;
+                if (this.enableDiscovery) {
+                    try {
+                        final ServerResponse controllerServerList =
+                                this.client.getControllerServerList(ServerRequest.getDefaultInstance());
+                        serverInfoGroup = ResolvedServerInfoGroup.builder()
+                                .addAll(controllerServerList.getNodeURIList()
+                                        .stream()
+                                        .map(node ->
+                                                new ResolvedServerInfo(
+                                                        new InetSocketAddress(node.getEndpoint(), node.getPort())))
+                                        .collect(Collectors.toList()))
+                                .build();
+                    } catch (StatusRuntimeException e) {
+                        log.warn("Failed to fetch controller addresses: ", e);
```

Comment:
The StatusRuntimeException provides a specific exception code which can be passed to the gRPC onError method, so I have retained this